arrows/core/serialize_metadata_csv.cxx

```diff
@@ -0,0 +1,251 @@
+            "\"Upper right corner point lon (EPSG:4326)\","
+            "\"Upper right corner point lat (EPSG:4326)\","
+            "\"Lower right corner point lon (EPSG:4326)\","
+            "\"Lower right corner point lat (EPSG:4326)\","
+            "\"Lower left corner point lon (EPSG:4326)\","
+            "\"Lower left corner point lat (EPSG:4326)\",";
+  }
+  else if( csv_field == kv::VITAL_META_VIDEO_URI )
+  {
+    // Exclude this field as it's long and redundant across frames
+    return;
+  }
+  else
+  {
+    // Quote all other data
+    fout << '"' << md_traits.tag_to_name( csv_field ) << "\",";
+  }
+}
+
+// ----------------------------------------------------------------------------
+serialize_metadata_csv
+::serialize_metadata_csv()
+  : d_{ new priv }
+{
+  attach_logger( "arrows.core.serialize_metadata_csv" );
+}
+
+// ----------------------------------------------------------------------------
+serialize_metadata_csv::
+~serialize_metadata_csv()
+{
+}
+
+// ----------------------------------------------------------------------------
+kv::metadata_map_sptr
+serialize_metadata_csv
+::load_( std::string const& filename ) const
+{
+  throw kv::file_write_exception( filename, "not implemented" );
+
+  kv::metadata_map::map_metadata_t metadata_map;
+
+  return std::make_shared< kv::simple_metadata_map >(
+    kv::simple_metadata_map( metadata_map ) );
+}
+
+// ----------------------------------------------------------------------------
+void
+serialize_metadata_csv
+::save_( std::string const& filename,
+         kv::metadata_map_sptr data ) const
+{
+  std::ofstream fout{ filename };
+
+  if( !fout )
+  {
+    VITAL_THROW( kv::file_write_exception, filename,
+                 "Insufficient permissions or moved file" );
+  }
+
+  // Accumulate the unique metadata IDs
+  std::set< kv::vital_metadata_tag > metadata_ids;
+  for( auto const& frame_data : data->metadata() )
+  {
+    for( auto const& metadata_packet : frame_data.second )
+    {
+      for( auto const& metadata_item : *metadata_packet )
+      {
+        kv::vital_metadata_tag type_id = metadata_item.first;
+        metadata_ids.insert( type_id );
+      }
+    }
+  }
+
+  // Write out the csv header
+  fout << "\"frame ID\",";
+  for( auto const& metadata_id : metadata_ids )
+  {
+    d_->write_csv_header( metadata_id, fout );
+  }
+  fout << std::endl;
+
+  for( auto const& frame_data : data->metadata() )
+  {
+    for( auto const& metadata_packet : frame_data.second )
+    {
+      // Write the frame number
+      fout << frame_data.first << ",";
+      for( auto const& metadata_id : metadata_ids )
+      {
+        if( metadata_packet->has( metadata_id ) )
+        {
+          d_->write_csv_item( metadata_packet->find( metadata_id ), fout );
+        }
+        // exclude this field from writing since it's internal and
+        // shared between all frames
+        else if( metadata_id != kv::VITAL_META_VIDEO_URI )
```

Comment:
Maybe refactor building `metadata_ids` so this is never in that set? That would allow removing this check and I think one or two others as well...